### PR TITLE
Multi Adapter support

### LINKED ISSUE #211: How to use multiple lora at the same time for text generation?
I have many lora models,For example: 
funny lora, Ultraman style lora ...
how to use them at the same time, and can adjust the weight between lora.
Similar to stable diffusion.

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Adds multi-adapter training and inference support. 
2. Fixes #211, #208, https://github.com/huggingface/peft/pull/133#issuecomment-1447801733, https://github.com/huggingface/peft/pull/133#issuecomment-1492990881 and https://gist.github.com/philschmid/821c5317d144250feef517aecd390b98

### Usage
1. While loading the first adapter via `PeftModel.from_pretrained`, you can give it a name using `**adapter_name**` parameter. Else the default adapter name `default` is used.
2. To load another adapter, use `**load_adapter()**` method of PeftModel, e.g., `model.load_adapter(peft_model_path, adapter_name)`
3. To switch between adapters, use `**set_adapter()**` method of PeftModel, e.g., `model.set_adapter(adapter_name)`
4. To disable adapters, use context manager `disable_adapter()`, e.g., `with model.disable_adapter()`
5. **Specific to LoRA method**: To merge and unload the current active adapter so that the lora weights are added to the base model weights and he injected models are removed to get back the transformers base model with lora weights added, use `merge_and_unload()` method, e.g.,  `model = model.merge_and_unload()`

```
from peft import PeftModel
from transformers import LlamaTokenizer, LlamaForCausalLM, GenerationConfig

model_name = "decapoda-research/llama-7b-hf"
tokenizer = LlamaTokenizer.from_pretrained(model_name)
model = LlamaForCausalLM.from_pretrained(
    model_name,
    load_in_8bit=True,
    device_map="auto",
    use_auth_token=True
)
model = PeftModel.from_pretrained(model, "tloen/alpaca-lora-7b", adapter_name="eng_alpaca")
model.load_adapter("22h/cabrita-lora-v0-1", adapter_name="portuguese_alpaca")

model.set_adapter("eng_alpaca")
instruction = "Tell me about alpacas."
print(evaluate(instruction))

model.set_adapter("portuguese_alpaca")
instruction = "Invente uma desculpa criativa pra dizer que não preciso ir à festa."
print(evaluate(instruction))

with model.disable_adapter():
    instruction = "Invente uma desculpa criativa pra dizer que não preciso ir à festa."
    print(evaluate(instruction))
```

Link to colab notebook: https://colab.research.google.com/drive/1vrVg8G7AIdCM9qpcfZya0B7OEsPSAxbO?usp=sharing

### CODE REVIEW

**Comment by Dentoty** [comment]
Would this work for Inference using LoRAs that were trained by using train_dreambooth.py ?

File: src/peft/tuners/lora.py [hunk 1]
```diff
@@ -328,224 +360,135 @@ def mark_only_lora_as_trainable(model: nn.Module, bias: str = "none") -> None:
...
-                    .transpose(-2, -1)
+        self.update_layer(adapter_name, r, lora_alpha, lora_dropout, init_lora_weights)
+        self.active_adapter = adapter_name
+
+    def merge(self):
+        if self.active_adapter not in self.lora_A.keys():
+            return
+        if not self.merge_weights:
+            warnings.warn("Nothing to merge. Set merge_weights to True to enable merging.")
+            return
```
Comment (younesbelkada) [inline src/peft/tuners/lora.py:441]: Maybe that could explain why the tests are failing, we should always call `merge_weights=False` (i.e. not test the case where`merge_weights=False`) --> or maybe call `.eval()` in case `merge_weigths=True`-

File: src/peft/tuners/lora.py [hunk 2]
```diff
@@ -328,224 +360,135 @@ def mark_only_lora_as_trainable(model: nn.Module, bias: str = "none") -> None:
...
         # Actual trainable parameters
         if r > 0:
-            self.lora_A = nn.Linear(in_features, r, bias=False)
-            self.lora_B = nn.Linear(r, out_features, bias=False)
-            self.scaling = self.lora_alpha / self.r
-            # Freezing the pre-trained weight matrix
-            self.weight.requires_grad = False
+            self.lora_A.update(nn.ModuleDict({adapter_name: nn.Linear(self.in_features, r, bias=False)}))
+            self.lora_B.update(nn.ModuleDict({adapter_name: nn.Linear(r, self.out_features, bias=False)}))
+            self.scaling[adapter_name] = lora_alpha / r
```
Comment (younesbelkada) [inline src/peft/tuners/lora.py:395]: I am wondering if this is a breaking change with respect to models that are already on the Hub? From what I have got you are replacing pure `nn.Linear` layers with a module dict. 
After thinking a bit, even if it's a breaking change, I believe we could provide a utility script to easily convert the `state_dict` from previous versions to the new ones

File: src/peft/utils/other.py [hunk 3]
```diff
@@ -113,11 +110,48 @@ def shift_tokens_right(input_ids: torch.Tensor, pad_token_id: int, decoder_start
...
+            else:
+                for param in target.parameters():
+                    param.requires_grad = True
+                setattr(parent, target_name, ModulesToSaveWrapper(target, adapter_name))
+
+
+def _set_adapter(model, adapter_name):
+    for module in model.modules():
+        if isinstance(module, ModulesToSaveWrapper):
+            module.active_adapter = adapter_name
```
Comment (younesbelkada) [inline src/peft/utils/other.py:154]: I see, this avoids the breaking change I believe, can you confirm? 🙏

File: src/peft/peft_model.py [hunk 4]
```diff
@@ -571,7 +645,8 @@ def forward(
         return_dict=None,
         **kwargs,
     ):
-        if not isinstance(self.peft_config, PromptLearningConfig):
+        peft_config = self.active_peft_config
```
Comment (younesbelkada) [inline src/peft/peft_model.py:648]: IMO it should be documented somewhere that the way retrieving the peft config has changed, now it's `active_peft_config` rather than `peft_config`

**Review by younesbelkada — COMMENTED** [review]
Awesome work! 🔥 This feature is really great
I left a couple of open questions, just to confirm if I have understood some changes correctly, my main concern was whether this introduced some breaking changes with respect to models that are on the Hub, but I think that you are dealing correctly with that using `ModulesToSaveWrapper`
IMO we just need to figure out why some tests are failing, they're related to merging layers for some models, I made a suggestion below (I think in the test we don't deal correctly with the case `merge_weights=False`.

File: src/peft/tuners/lora.py (hunk 2 above)
Comment (pacman100) [reply src/peft/tuners/lora.py:395]: This isn't a breaking change, as the example colab in description already uses peft checkpoints from hub

Comment (pacman100) [reply src/peft/tuners/lora.py:395]: I handle loading and saving utilities to be backward compatible and hence avoid breaking changes

Comment (younesbelkada) [reply src/peft/tuners/lora.py:395]: Awesome! That's great then

File: src/peft/utils/other.py (hunk 3 above)
Comment (pacman100) [reply src/peft/utils/other.py:154]: This avoids breaking changes when there are additional trainable layers such as `classfier head/regression head` on top of the model for tasks like AutoModelForSequenceClassification or TRL reward model head ... these are also saved along with the adapter weights and if each checkpoint has its own additional trainable layers, this makes sure that they are properly being called.

Comment (younesbelkada) [reply src/peft/utils/other.py:154]: Awesome, this is great then!

**Comment by pacman100** [comment]
Hello @younesbelkada, in my tests wherein I ran all the example notebooks and scripts, the only breaking change is loading peft checkpoints from hub which used MergedLinear.

Why remove support for MergedLinear?
1. The [AdaLoRA](https://arxiv.org/abs/2303.10512) and [MAM](https://arxiv.org/pdf/2110.04366.pdf) papers have shown that the best performance is achieved by incorporating LoRA adapters in FFN in addition to MHA and there is no specific reason to exclude the `key` layer.  So, I would gradually see more people using custom target modules or AdaLoRA (Adaptive LoRA). This makes mergedLinear a bit redundant.
2. Maintainance and complexity of MergedLinear are high compared to the functionality. Post removing this, the code is clean and easy to maintain and doesn't disable any model from supporting LoRA.

Other changes:
1. Now `model.eval()` no longer merges the weights of the lora into base model. It is now achieved by `model.merge_adapter()` and `model.unmerge_adapter()` for unmerging. This is based on user feedback that something explicit like this is better than `model.eval()` which people normally won't associate with merging lora adapters as they already have different purpose in pytorch. Let me know if this could lead to confusion and if we should support that too with decrecation warnings

**Comment by totaltube** [comment]
Tested in production - works great!

**Review by younesbelkada — COMMENTED** [review]
Awesome work as always! 🚀 
Let's just fix the T5 failing test and we should be good to merge

**Comment by lxe** [comment]
Great timing on this!

**Comment by Jingsong-Yan** [comment]
`disable_adapter()` doesn't seem to work in my tests.

**Comment by llllllim** [comment]
How can I just upload the lora and remain the base model？I want to reduce memory usage

**Comment by HyperdustLab** [comment]
this is nice
